File: python/불량사용자.py
def check(b, u):
    for x in range(len(b)):
        if b[x] != u[x] and b[x] != '*':
            return False
    return True

def solution(user_id, banned_id):
    
    s = []
    from itertools import permutations
    a = list(permutations(user_id, len(banned_id)))

    for i in range(len(a)):
        user = a[i]
        j = 0
        temp = []
        for k in range(len(user)):
            if len(banned_id[j]) != len(user[k]):
                break
            if not check(banned_id[j], user[k]):
                break
            else:
                temp.append(user[k])
                j += 1
            if len(temp) == len(banned_id):
                if sorted(temp) not in s:
                    s.append(sorted(temp))
                break
            if j >= len(banned_id):
                break
    return len(s)

print(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],["fr*d*", "abc1**"]))
print(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],["*rodo", "*rodo", "******"]))
print(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],["fr*d*", "*rodo", "******", "******"]))
print(solution(["12345", "13579"],["1****", "12***"]))

File: JavaScript/불량사용자.js
const getAnswer = (uid, bid, visited) => {
    if (bid.length === 0) {
        if (!visited.includes(JSON.stringify(uid))){
            visited.push(JSON.stringify(uid));
            return 1;
        } else {
            return 0;
        }
    }
    let count = 0;
    for (let i = 0; i < uid.length; i++) {
        for (let j = 0; j < bid.length; j++) {
            if (uid[i].length !== bid[j].length) continue;
            let equal = true;
            for (let k = 0; k < uid[i].length; k++) {
                if (bid[j][k] !== '*' && uid[i][k] !== bid[j][k]) {
                    equal = false;
                    break;
                }
            }
            if (!equal) continue;
            const nextUid = uid.slice();
            const nextBid = bid.slice();
            nextUid.splice(i, 1);
            nextBid.splice(j, 1);
            if (!visited.includes(JSON.stringify(nextUid) + JSON.stringify(nextBid))) {
                visited.push(JSON.stringify(nextUid) + JSON.stringify(nextBid));
                count += getAnswer(nextUid, nextBid, visited);
            }
        }
    }
    return count;
}

function solution(user_id, banned_id) {
    return getAnswer(user_id, banned_id, []);
}

//console.log(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"], ["*rodo", "*rodo", "******"]))
console.log(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"], ["fr*d*", "*rodo", "******", "******"]))
//console.log(solution(["aaaaaaaa", "bbbbbbbb", "cccccccc", "dddddddd", "eeeeeeee", "ffffffff", "gggggggg", "hhhhhhhh"], ["********", "********", "********", "********", "********", "********", "********", "********"]))